### PR TITLE
Enable lowering all float types to tosa

### PR DESCRIPTION
**Pull request by josel-amd**
Allow lowering of all floating point types available in onnx-mlir to tosa, since tosa now allows arbitrary floating point types when no specific mode is used (see https://github.com/Xilinx/llvm-project/pull/175).

### CODE REVIEW

File: test/mlir/conversion/onnx_to_tosa/Flow/fp8.mlir [hunk 1]
```diff
@@ -0,0 +1,36 @@
+// RUN: onnx-mlir-opt --shape-inference --convert-onnx-to-tosa %s -split-input-file | FileCheck %s
```
Comment (josel-amd) [inline test/mlir/conversion/onnx_to_tosa/Flow/fp8.mlir:1]: I'm not completely sure about the test organization here. It seems that every operator checks all the types, is there an order?

Comment (mgehre-amd) [reply test/mlir/conversion/onnx_to_tosa/Flow/fp8.mlir:1]: I think doing one test for place where `isTOSAFloat` was used should be fine.

File: src/Conversion/ONNXToTOSA/ConvertONNXToTOSA.cpp [hunk 2]
```diff
@@ -120,8 +125,8 @@ void FrontendToTosaLoweringPass::runOnOperation() {
   // conversion failures. Quantized types are not supported right now.
   TypeConverter typeConverter;
   typeConverter.addConversion([](Type type) -> std::optional<Type> {
-    if (isTOSAInt(type) || isTOSAFloat(type) || type.isa<NoneType>() ||
-        isTOSABool(type))
+    if (isTOSAInt(type) || isTOSAFloat(type) || isFloat8(type) ||
```
[the remark this answers is not in the document]
Comment (mgehre-amd) [reply src/Conversion/ONNXToTOSA/ConvertONNXToTOSA.cpp:128]: Yes, I don't think we need to decide here. We still have the input verifier to make the decision of what the BE supports, and in the future we might also get an ONNX level input verifier.

[the remark this answers is not in the document]
Comment (josel-amd) [reply src/Conversion/ONNXToTOSA/ConvertONNXToTOSA.cpp:128]: Alright

File: test/mlir/conversion/onnx_to_tosa/Flow/fp8.mlir [hunk 3]
```diff
@@ -0,0 +1,36 @@
+// RUN: onnx-mlir-opt --convert-onnx-to-tosa %s -split-input-file | FileCheck %s
+
```
Comment (mgehre-amd) [inline test/mlir/conversion/onnx_to_tosa/Flow/fp8.mlir:2]: Can you please add a test in the `resize.mlir` / transpose.mlir and `elementwise.mlir` files corresponding to the lines you changed?

File: src/Dialect/ONNX/ONNXOps.td.inc [hunk 4]
```diff
@@ -7241,7 +7241,7 @@ def ONNXResizeOp:ONNX_Op<"Resize",
   ```
   if input \\"sizes\\" is not specified.
   }];
-  let arguments = (ins AnyTypeOf<[TensorOf<[UI8]>, TensorOf<[UI16]>, TensorOf<[UI32]>, TensorOf<[UI64]>, TensorOf<[I8]>, TensorOf<[I16]>, TensorOf<[I32]>, TensorOf<[I64]>, TensorOf<[BF16]>, TensorOf<[F16]>, TensorOf<[F32]>, TensorOf<[F64]>, TensorOf<[StringType]>, TensorOf<[I1]>, TensorOf<[Complex<F32>]>, TensorOf<[Complex<F64>]>]>:$X,
```
Comment (mgehre-amd) [inline src/Dialect/ONNX/ONNXOps.td.inc:7244]: Please revert changes in this file

File: test/mlir/conversion/onnx_to_tosa/Tensor/Resize.mlir [hunk 5]
```diff
@@ -240,3 +240,21 @@ func.func @test_resize_cubic_disallowed(%arg0: tensor<1x1x2x4xf32>) -> tensor<1x
 // CHECK-LABEL:  func.func @test_resize_cubic_disallowed
 // CHECK-LABEL:  onnx.Resize
 }
+
```
Comment (mgehre-amd) [inline test/mlir/conversion/onnx_to_tosa/Tensor/Resize.mlir:243]:
```suggestion
```

File: test/mlir/conversion/onnx_to_tosa/Tensor/Resize.mlir [hunk 6]
```diff
@@ -240,3 +240,21 @@ func.func @test_resize_cubic_disallowed(%arg0: tensor<1x1x2x4xf32>) -> tensor<1x
...
+// -----
+
+func.func @test_resize_half_pixel_nearest_floor_downsample_axis_one_fp8(%arg0: tensor<1x1x1x12xf8E4M3FN>) -> tensor<1x1x1x6xf8E4M3FN> {
+    %0 = "onnx.NoValue"() {value} : () -> none
+    %1 = "onnx.Constant"() {value = dense<[6]> : tensor<1xi64>} : () -> tensor<1xi64>
+    %2 = "onnx.Resize"(%arg0, %0, %0, %1) {axes = [3], coordinate_transformation_mode = "half_pixel", cubic_coeff_a = -7.500000e-01 : f32, exclude_outside = 0 : si64, extrapolation_value = 0.000000e+00 : f32, mode = "nearest", nearest_mode = "floor"} : (tensor<1x1x1x12xf8E4M3FN>, none, none, tensor<1xi64>) -> tensor<1x1x1x6xf8E4M3FN>
+    return %2 : tensor<1x1x1x6xf8E4M3FN>
+// CHECK-LABEL: @test_resize_half_pixel_nearest_floor_downsample_axis_one_fp8
+// CHECK-SAME: ([[PARAM_0_:%.+]]: tensor<1x1x1x12xf8E4M3FN>) -> tensor<1x1x1x6xf8E4M3FN> {
+// CECK-NEXT: [[VAR_0_:%.+]] = "tosa.const"() <{value = dense<[0, 2, 3, 1]> : tensor<4xi32>}> : () -> tensor<4xi32>
```
Comment (mgehre-amd) [inline test/mlir/conversion/onnx_to_tosa/Tensor/Resize.mlir:254]:
```suggestion
// CHECK-NEXT: [[VAR_0_:%.+]] = "tosa.const"() <{value = dense<[0, 2, 3, 1]> : tensor<4xi32>}> : () -> tensor<4xi32>
```
and below

File: test/mlir/conversion/onnx_to_tosa/Flow/fp8.mlir (hunk 3 above)
Comment (TinaAMD) [reply test/mlir/conversion/onnx_to_tosa/Flow/fp8.mlir]: What kind of testing should we add here? I think for especially most of the elementwise ops there is only tests with f32. Do would want to add all kinds of ops with their allowed types (which are now lowered or were already lowered but not tested)? That would mean tests for 3-5 additional float types per op for around 18 ops.

Comment (TinaAMD) [reply test/mlir/conversion/onnx_to_tosa/Flow/fp8.mlir]: Discussed offline: Added tests for some elementwise ops which use the changed functions.

**Review by mgehre-amd — APPROVED** [review]

**Comment by mgehre-amd** [comment]
Thanks, looks good!

**Review by TinaAMD — APPROVED** [review]

**Pull request merged**